src/python/pants/backend/python/macros/poetry_requirements.py

```diff
@@ -167,62 +167,92 @@ def non_pants_project_abs_path(self, path: str) -> Path | None:
+        f"{')' if (' and ' in markers_lookup or ' or ' in markers_lookup) else ''}"
+        f"{' and ' if python_lookup and markers_lookup else ''}"
+        f"{python_lookup}"
+    )
+
+
 def handle_dict_attr(
-    proj_name: str, attributes: dict[str, str], pyproject_toml: PyProjectToml
+    proj_name: str, attributes: Mapping[str, Sequence[str] | str], pyproject_toml: PyProjectToml
 ) -> str | None:
-    def produce_match(sep: str, feat: Optional[str]) -> str:
-        return f"{sep}{feat}" if feat else ""
+    base = ""
+    fp = str(pyproject_toml.toml_relpath)
+
+    extras_lookup = attributes.get("extras")
+    if type(extras_lookup) == list:
+        extras_str = f"[{','.join(extras_lookup)}]"
+    elif type(extras_lookup) == str:
+        extras_str = f"{extras_lookup}"
+    else:
+        extras_str = ""
 
     git_lookup = attributes.get("git")
     if git_lookup is not None:
         rev_lookup = produce_match("#", attributes.get("rev"))
         branch_lookup = produce_match("@", attributes.get("branch"))
         tag_lookup = produce_match("@", attributes.get("tag"))
 
-        return f"{proj_name} @ git+{git_lookup}{tag_lookup}{branch_lookup}{rev_lookup}"
-
-    version_lookup = attributes.get("version")
+        base = f"{proj_name}{extras_str} @ git+{git_lookup}{tag_lookup}{branch_lookup}{rev_lookup}"
 
     path_lookup = attributes.get("path")
     if path_lookup is not None:
         non_pants_project_abs_path = pyproject_toml.non_pants_project_abs_path(path_lookup)
         if non_pants_project_abs_path:
-            return f"{proj_name} @ file://{non_pants_project_abs_path}"
-        # An internal path will be handled by normal Pants dependencies and dependency inference;
-        # i.e.: it never represents a third party requirement.
-        return None
+            base = f"{proj_name}{extras_str} @ file://{non_pants_project_abs_path}"
+        else:
+            # An internal path will be handled by normal Pants dependencies and dependency inference;
+            # i.e.: it never represents a third party requirement.
+            return None
 
     url_lookup = attributes.get("url")
     if url_lookup is not None:
-        return f"{proj_name} @ {url_lookup}"
+        base = f"{proj_name}{extras_str} @ {url_lookup}"
 
+    version_lookup = attributes.get("version")
     if version_lookup is not None:
-        markers_lookup = produce_match(";", attributes.get("markers"))
-        fp = str(pyproject_toml.toml_relpath)
-        python_lookup = parse_python_constraint(attributes.get("python"), fp)
-        version_parsed = parse_str_version(proj_name, version_lookup, fp)
-        return (
-            f"{version_parsed}"
-            f"{markers_lookup}"
-            f"{' and ' if python_lookup and markers_lookup else (';' if python_lookup else '')}"
-            f"{python_lookup}"
+        base = parse_str_version(proj_name, version_lookup, fp, extras_str)
+
+    if len(base) == 0:
+        raise AssertionError(
```

Comment:
Assertion because you expect this to never actually happen in the wild? If so, that's the correct use of AssertionError. Otherwise, you probably want ValueError and to mention where this error is coming from (poetry macro).

Reply:
The only case in which that happens is where you have a malformed requirement? I think ValueError does make more sense.

Reply:
Yeah, assertions are meant to check issues in our own code. For example, MyPy thinks the type is `str | list[str]`. You are confident it is going to be `str`. Rather than `# type: ignore` or `cast()`, often even better is to use `assert isinstance(x, str)`. If you were wrong about your assumption, Python will error.

Assertions shouldn't be used when the user could feasibly make the mistake.